[libc++] Implement C++20 atomic_ref

dalg24 opened the pull request:
Implement `atomic_ref` class template by reusing `atomic_base_impl`.
Based on the work from https://reviews.llvm.org/D72240

dalg24 committed:
[libc++][atomic_ref] Refactor atomic_base_impl class and friends

so it can be reused to implement atomic_ref.
Salvaged/adapted from https://reviews.llvm.org/D72240

Co-Authored-By: Benjamin Trapani

dalg24 committed:
[libc++][atomic_ref] Enable atomic load/exchange for non-default constructible types

dalg24 committed:
[libc++][atomic_ref] Add _LIBCPP_CHECK_WAIT_MEMORY_ORDER macro

dalg24 committed:
[libc++][atomic_ref] Implement C++20 atomic_ref

dalg24 committed:
[libc++][atomic_ref] Add tests for atomic_ref

dalg24 committed:
[libc++][atomic_ref] Fix shadow warning

jyknight commented:
It's important to note that this PR deletes the codepath in cxx_atomic_impl.h which uses the `__c11_*` builtins and C `_Atomic(T)` to implement `std::atomic`.

That's a wonderful simplification, and I would _love_ to see it done. BUT, it changes the ABI for `std::atomic`. It both makes it incompatible with previous libc++, and with Clang's C11 `_Atomic(T)` type.

Unfortunately, the current situation is a total mess.

### Mess 1: _Atomic ABI

Clang's ABI for _Atomic(T) is complex: it attempts to round up both size and alignment of T, but only up to a per-platform limit, to ensure that a lock-free atomic can be used if possible. That is: given `struct X { char n[3] };`, on some platforms, `sizeof(std::atomic<X>)` and `alignof(std::atomic<X>)`, will be 4 and 4, but on others, it'll be 3 and 1. See the `MaxAtomicPromoteWidth` values set by clang/lib/Basic/Targets/*. 

Which means C11 atomics in Clang are ABI-incompatible with GCC. GCC never adds padding, but instead simply increases alignment for objects of size 1, 2, 4, 8, 16 to match their size. Quite unfortunate situation, since usually GCC and Clang are ABI-compatible for C code. 

So: the ABI for C11 _Atomic differs between Clang and GCC. The ABI for libc++'s std::atomic _also_ changes depending on which compiler you build with (since it changes which underlying implementation strategy is used). It matches C11 _Atomic ABI when built on Clang, but when built on GCC, uses an ABI incompatible with both.

GCC's libstdc++ does have a consistent ABI between Clang and GCC, because it always uses GCC's rule, even when built with Clang. But that means it doesn't match C11 _Atomic on Clang, only on GCC.

Making everything worse is that, as of C++23's addition of <stdatomic.h>,  expectation of ABI-compatibility between C++ `std::atomic<T>` and C11 `_Atomic(T)` is heightened.

This is something that I think we ought to fix -- but it would require an ABI break. A long time ago, I attempted to bring this up as something the psABI documents should _define_, but didn't manage to push hard enough to get it actually done. My expectation is that if the documents did specify it, they would specify what GCC does. Clang's option is, in principle, nice from a QoI standpoint (it allows more types to be lock-free), but in practice is simply unnecessary complexity.

### Mess 2: libc++ atomic configuration options

OK, so, that's a huge minefield. But for atomic_ref, we can actually ignore all of the above complexity, since we cannot change the layout of the underlying type. Thus: we can simply implement atomic_ref directly on the `__atomic_*` builtins, without going through libc++'s `__cxx_atomic_*` abstraction layer -- leaving fixing that mess for another day. The abstraction layer really doesn't get us anything!

Well, _except_ that libc++ has a _third_ implementation strategy that std::atomic can use. Namely, 21450545d144ed85cb5bc15ab3e15f9a6583af40 added a custom locked impl under `_LIBCPP_ATOMIC_ONLY_USE_BUILTINS`.

IMO, that is effectively superfluous and incompatible with the goal of C and C++ equivalency, and I don't really understand the rationale for adding it in the first place. If someone wants to use locked-atomics in a freestanding implementation, they should simply provide the requisite libatomic libcalls for their environment, end of story.

So, I think we should remove this third option. Or, at least, ignore it when implementing atomic_ref, and do just use the `__atomic_*` builtins directly.

dalg24 committed:
[libc++][atomic_ref] Reimplement atomic_ref in terms of the GCC __atomic builtins

dalg24 committed:
[libc++][atomic_ref] Revert all changes to __atomic/cxx_atomic_impl.h

dalg24 commented:
* Fixed the format (was using the wrong clang-format version...)
* Fixed some shadow warnings in the test
* Followed James' suggestion in https://github.com/llvm/llvm-project/pull/76647#issuecomment-1872988253 and dropped all the changes to `atomic_base_impl` and implemented `atomic_ref` directly in terms of the GCC `__atomic_*` builtins
* Haven't implemented `wait`, `notify_one`, nor `notify_all` yet

dalg24 committed:
[libc++][atomic_ref] move __to_gcc_[failure_]order to its own header file

dalg24 committed:
[libc++][atomic_ref] add header to module map

dalg24 committed:
[libc++][atomic_ref] fixup atomic_{add,sub} for floating points

dalg24 committed:
[libc++][atomic_ref] cleanup header includes

dalg24 committed:
[libc++][atomic_ref] annotate wait, notify_one, and notify_all as expected failures

huixie90 reviewed (commented):
I have not yet reviewed the code. Just to bring up some issues we encountered previously:

there are some weird downstream setup. IIRC, there is Fushia where it is Linux but without gcc installed (thus cannot -latomic).

there is also Windows clang-cl which lacks of some builtins. 

On Mac, do we have the gcc builtin at all? At least, on my old laptop all the code path are going through the c11 branch. 

I’d also like to mention that I am working on several atomic related things, including making types with padding bytes in between member and/or after the tail works with atomic. (btw, does your atomic_ref compare_exchange_strong works for those types?) I don’t think we are allowed to touch the padding bits since atomic_ref does not own the object 

Also the current c11_compare_exchange is broken for types whose size is not power of two. And I am also working on fixing that. 

and I don’t think breaking existing std::atomic ABI is feasible. even though we could Abi tag std::atomic, there is no way we could tag user defined classes that contain std::atomic and appear in Abi boundary.

jyknight commented:
> there are some weird downstream setup. IIRC, there is Fushia where it is Linux but without gcc installed (thus cannot -latomic).

That doesn't affect whether you can use the `__atomic_*` builtin functions.

> there is also Windows clang-cl which lacks of some builtins.

Fortunately, it doesn't lack these.

> On Mac, do we have the gcc builtin at all? At least, on my old laptop all the code path are going through the c11 branch.

Yep, they exist on all Clang.

> I’d also like to mention that I am working on several atomic related things, including making types with padding bytes in between member and/or after the tail works with atomic. (btw, does your atomic_ref compare_exchange_strong works for those types?) I don’t think we are allowed to touch the padding bits since atomic_ref does not own the object

This PR indeed does not implement that part of the spec yet. That'll need to wait until `__builtin_clear_padding` is implemented. I think it's fine to implement this type without that requirement, initially. (note that when the time comes, it will need to use a different implementation than std::atomic (and it'll be extremely inefficient, since it cannot assume the padding is cleared to start with!)

> and I don’t think breaking existing std::atomic ABI is feasible. even though we could Abi tag std::atomic, there is no way we could tag user defined classes that contain std::atomic and appear in Abi boundary.

Recall that long-term it's not only std::atomic whose ABI we need to change, but also C11 _Atomic. However, this PR no longer touches it, so that's moot for the purposes of this PR.

huixie90 commented:
> > there are some weird downstream setup. IIRC, there is Fushia where it is Linux but without gcc installed (thus cannot -latomic).
> 
> That doesn't affect whether you can use the `__atomic_*` builtin functions.
> 
> > there is also Windows clang-cl which lacks of some builtins.
> 
> Fortunately, it doesn't lack these.
> 
> > On Mac, do we have the gcc builtin at all? At least, on my old laptop all the code path are going through the c11 branch.
> 
> Yep, they exist on all Clang.
> 
> > I’d also like to mention that I am working on several atomic related things, including making types with padding bytes in between member and/or after the tail works with atomic. (btw, does your atomic_ref compare_exchange_strong works for those types?) I don’t think we are allowed to touch the padding bits since atomic_ref does not own the object
> 
> This PR indeed does not implement that part of the spec yet. That'll need to wait until `__builtin_clear_padding` is implemented. I think it's fine to implement this type without that requirement, initially. (note that when the time comes, it will need to use a different implementation than std::atomic (and it'll be extremely inefficient, since it cannot assume the padding is cleared to start with!)
> 
> > and I don’t think breaking existing std::atomic ABI is feasible. even though we could Abi tag std::atomic, there is no way we could tag user defined classes that contain std::atomic and appear in Abi boundary.
> 
> Recall that long-term it's not only std::atomic whose ABI we need to change, but also C11 _Atomic. However, this PR no longer touches it, so that's moot for the purposes of this PR.

just to put some related in progress work I have:

- __builtin_clear_padding https://github.com/llvm/llvm-project/pull/75371
- atomic_base_impl  clear padding https://github.com/llvm/llvm-project/pull/76180
- bug in std::atomic with not-power-of-2 size  (see the comments I put in the end) https://github.com/llvm/llvm-project/issues/30023

I think your patch is orthogonal to all these bugs we have so we can try to land this one first.

dalg24 committed:
[libc++][atomic_ref] fix generic-modules build

dalg24 committed:
[libc++][atomic_ref] uncomment `using std::atomic_ref;` in atomic module

dalg24 committed:
[libc++][atomic_ref] add missing header include for uintptr_t caught by hardening mode CI builds

dalg24 committed:
[libc++][atomic_ref] Attempt to use __cxx_atomic_wait

dalg24 committed:
[libc++][atomic_ref] Fiddle with hardening-mode annotations in preconditions tests

dalg24 committed:
[libc++][atomic_ref] Fix cxx03 CI build

dalg24 committed:
Revert "[libc++][atomic_ref] Fiddle with hardening-mode annotations in preconditions tests"

This reverts commit a132470db3beae515ad47afef2f27e9d65e7303e.

dalg24 committed:
[libc++][atomic_ref] Robust against ADL lookup

ldionne commented:
> IMO, that is effectively superfluous and incompatible with the goal of C and C++ equivalency, and I don't really understand the rationale for adding it in the first place. If someone wants to use locked-atomics in a freestanding implementation, they should simply provide the requisite libatomic libcalls for their environment, end of story.
> 
> So, I think we should remove this third option. Or, at least, ignore it when implementing atomic_ref, and do just use the `__atomic_*` builtins directly.

I would support that too. https://reviews.llvm.org/D56913 happened a long time ago, and if it were to be done again, I might push back against that approach as being too ad-hoc. I'm not even certain we use this anywhere, and I'd argue that Freestanding does not equate there being no builtins. So +1 from me for removing it, at least at first glance.

ldionne commented on libcxx/test/std/atomics/atomics.ref/convert.pass.cpp:
```diff
@@ -0,0 +1,47 @@
+  T x(T(1));
+  std::atomic_ref<T> a(x);
+
+  assert(a == T(1));
```
I think the best we can do here would be:

```c++
T converted = a;
assert(converted == T(1));
```

dalg24 replied:
Done

ldionne commented on libcxx/test/std/atomics/atomics.ref/convert.pass.cpp:
```diff
@@ -0,0 +1,47 @@
+}
+
+void test() {
+  test_convert<int>();
+
+  test_convert<float>();
+
+  test_convert<int*>();
+
```
I would suggest doing `test_convert<int>(3)`, `test_convert<int*>(nullptr)`, etc..

That way you don't need the sneaky casting in `T x(T(1))`.

dalg24 replied:
Done for that file.  Still using that pattern in other places when I need multiple value representations.  Could not think of an elegant way to avoid it.

ldionne commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,276 @@
+  _LIBCPP_HIDE_FROM_ABI _Tp exchange(_Tp __desired, memory_order __order = memory_order::seq_cst) const noexcept {
+    alignas(_Tp) unsigned char __mem[sizeof(_Tp)];
+    auto* __ret = reinterpret_cast<_Tp*>(__mem);
+    __atomic_exchange(__ptr_, std::addressof(__desired), __ret, __to_gcc_order(__order));
```
Here and everywhere else:
```suggestion
    __atomic_exchange(__ptr_, std::addressof(__desired), __ret, std::__to_gcc_order(__order));
```

We do this for all free function calls to avoid ADL kicking in. In this case though ADL can't be a problem since we know the argument is a `memory_order` and not some user defined thing, but we should still do it for consistency.

dalg24 replied:
Done

ldionne reviewed (changes requested):
Thanks a lot for the patch! The review doesn't go over the implementation in detail since I think we need to consider the overall state of `atomic_wait` primitives and GCC vs Clang atomics before we commit to an implementation strategy here.

dalg24 committed:
[libc++][atomic_ref] Qualify all __to_gcc_order calls

dalg24 committed:
[libc++][atomic_ref] Simplify inheritance from __atomic_ref_base

dalg24 committed:
[libc++][atomic_ref] Prefer snake_case

dalg24 committed:
[libc++][atomic_ref] Use proper _LIBCPP_ASSERT_* macro

dalg24 committed:
[libc++][atomic_ref] Prefix assertion messages with "atomic_ref:"

dalg24 committed:
[libc++][atomic_ref] Fix XFAIL annotations

dalg24 committed:
[libc++][atomic_ref] Fixup conversion test

dalg24 committed:
[libc++][atomic_ref] Drop test that the ctor is explicit since there is already a check

dalg24 committed:
[libc++][atomic_ref] Fixup do not use -Wno-ctad-maybe-unsupported in CTAD tests

dalg24 committed:
[libc++][atomic_ref] Rename {type -> requires-trivially-copyable}.verify.cpp

dalg24 committed:
[libc++][atomic_ref] Const the atomic_refs in the tests

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,276 @@
+    _LIBCPP_ASSERT_UNCATEGORIZED(
+        __order == memory_order::relaxed || __order == memory_order::consume || __order == memory_order::acquire ||
+            __order == memory_order::seq_cst,
+        "memory order argument to atomic load operation is invalid");
```
Done

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,276 @@
+
+  _LIBCPP_HIDE_FROM_ABI _Tp load(memory_order __order = memory_order::seq_cst) const noexcept
+      _LIBCPP_CHECK_LOAD_MEMORY_ORDER(__order) {
+    _LIBCPP_ASSERT_UNCATEGORIZED(
```
Done

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,276 @@
+
+#if _LIBCPP_STD_VER >= 20
+
+template <class _Tp, bool = is_integral_v<_Tp> && !is_same_v<_Tp, bool>, bool = is_floating_point_v<_Tp>>
```
Done

dalg24 commented on libcxx/test/libcxx/atomics/atomics.ref/assert.load.pass.cpp:
```diff
@@ -0,0 +1,60 @@
+// REQUIRES: has-unix-headers
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: libcpp-hardening-mode=none
+// XFAIL: availability-verbose_abort-missing
```
Done

dalg24 commented on libcxx/test/std/atomics/atomics.ref/ctor.explicit.verify.cpp:
```diff
@@ -0,0 +1,34 @@
+//
```
Removed.

dalg24 commented on libcxx/test/std/atomics/atomics.ref/deduction.pass.cpp:
```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// ADDITIONAL_COMPILE_FLAGS: -Wno-ctad-maybe-unsupported
```
I dropped the additional compile flag.  Thanks.

dalg24 commented on libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp:
```diff
@@ -0,0 +1,47 @@
+#include "test_macros.h"
+
+template <typename T>
+void checkAlwaysLockFree(std::atomic_ref<T> a) {
```
Done

dalg24 commented on libcxx/test/std/atomics/atomics.ref/load.pass.cpp:
```diff
@@ -0,0 +1,48 @@
+template <typename T>
+void test_load() {
+  T x(T(1));
+  std::atomic_ref<T> a(x);
```
Done here and everywhere else.

dalg24 commented on libcxx/test/std/atomics/atomics.ref/type.verify.cpp:
```diff
@@ -0,0 +1,26 @@
+//
```
Renamed.

dalg24 commented on libcxx/test/std/atomics/atomics.ref/required_alignment.pass.cpp:
```diff
@@ -0,0 +1,34 @@
+template <typename T>
+void checkRequiredAlignment() {
+  static_assert(std::atomic_ref<T>::required_alignment >= alignof(T));
+}
```
Done.

dalg24 committed:
[libc++][atomic_ref] Let required_alignement test check both at runtime and compile time

dalg24 committed:
[libc++][atomic_ref] Add a release note and update the status page

dalg24 commented on libcxx/include/CMakeLists.txt:
```diff
@@ -236,6 +236,7 @@ set(files
   __atomic/atomic_flag.h
```
Done

dalg24 committed:
[libc++][atomic_ref] Update generated files

dalg24 committed:
[libc++][atomic_ref] Fix clang-tidy warnings modernize-use-nullptr

dalg24 committed:
[libc++][atomic_ref] Use std::same_as<T> trick

dalg24 commented on libcxx/test/std/atomics/atomics.ref/load.pass.cpp:
```diff
@@ -0,0 +1,48 @@
+  T x(T(1));
+  std::atomic_ref<T> a(x);
+
+  assert(a.load() == T(1));
```
Done

dalg24 committed:
[libc++][atomic_ref] Forgot to update assertion msg in tests

dalg24 committed:
[libc++][atomic_ref] Fix test compare_exchange_{strong,weak} assertion msg

dalg24 committed:
[libc++][atomic_ref] Forgot to use addressof in __atomic_ref_base constructor

dalg24 committed:
[libc++][atomic_ref] Prepare the code for clearing padding bits

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,335 @@
+        __order == memory_order::relaxed || __order == memory_order::consume || __order == memory_order::acquire ||
+            __order == memory_order::seq_cst,
+        "atomic_ref: memory order argument to atomic wait operation is invalid");
+    std::__cxx_atomic_wait(__ptr_, __old, __order);
```
I don't think this is going to work, at least on Mac.
`__cxx_atomic_wait` is assuming the first argument being `__cxx_atomic_impl<Tp>*` and it is calling some functions from the abstraction layer (which you are not using). so on Mac, this is going to call the functions that assumes `_Atomic` in various places

huixie90 commented on libcxx/include/__atomic/atomic_sync.h:
```diff
@@ -27,6 +30,21 @@
+template <class _Tp>
+struct __is_cxx_atomic_impl<__cxx_atomic_impl<_Tp> > : true_type {};
+
+template <class _Tp>
+_LIBCPP_HIDE_FROM_ABI __enable_if_t<!__is_cxx_atomic_impl<__remove_cv_t<_Tp> >::value, _Tp>
+__cxx_atomic_load(_Tp* __ptr, memory_order __order) _NOEXCEPT {
```
I think this is workaround for the `cxx_atomic_wait` not working? to be honest, I think this is making this header even more complicated. I would rather do it properly

dalg24 replied:
This is a hack to serve as a base for discussion of how we actually want to handle notify/wait.

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,335 @@
+
+  _LIBCPP_HIDE_FROM_ABI static _Tp* __clear_padding(_Tp& __val) noexcept {
+    _Tp* __ptr = std::addressof(__val);
+#  if __has_builtin(__builtin_clear_padding)
+    if constexpr (__might_have_padding) {
```
You're right

dalg24 committed:
[libc++][atomic_ref] Avoid sneaky casting `T x(T(1))``

dalg24 committed:
[libc++][atomic_ref] Check more types for is_always_lock_free

dalg24 committed:
[libc++][atomic_ref] Use fake minimally aligned pointer in is_lock_free()

dalg24 committed:
[libc++][atomic_ref] Include missing <cstring> header for memcmp and memcpy

dalg24 commented on libcxx/test/std/atomics/atomics.ref/is_always_lock_free.pass.cpp:
```diff
@@ -0,0 +1,47 @@
+  int* p = &i;
+  checkAlwaysLockFree(std::atomic_ref<int*>(p));
+
+  struct X {
```
Done.  Had to fiddle with the definition of `is_[always]_lock_free`

ldionne commented on libcxx/test/libcxx/atomics/atomics.ref/assert.compare_exchange_strong.pass.cpp:
```diff
@@ -0,0 +1,63 @@
+//===----------------------------------------------------------------------===//
+// REQUIRES: has-unix-headers
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: libcpp-hardening-mode=none
```
This should be `UNSUPPORTED: libcpp-hardening-mode=none || libcpp-hardening-mode=fast`

dalg24 replied:
Done

ldionne reviewed (commented):
I just met with @huixie90 and @dalg24 to chat about this patch. We had three key take-aways:

1. It seems fine to use the GCC atomics unconditionally to implement `atomic_ref`. There is no question of potentially using `_Atomic(T)` anyway, since we don't control the storage of the `T`.
2. We agreed that `_LIBCPP_ATOMIC_ONLY_USE_BUILTINS` seemed undesirable, as brought up by @jyknight. Indeed, the goal of that setting was to decouple libc++ from libraries like compiler-rt and libatomic in Freestanding mode. We didn't have a clear understanding of goals and non-goals of Freestanding back then, but nowadays we do have a better understanding that removing all dependencies of libc++ in Freestanding is a non-goal. We should still be able to depend on builtins like those defined in compiler-rt for implementing our atomic operations under Freestanding. Hence, it makes sense to remove this setting (which should not have wide usage anyway). I'll get started on a patch to remove it. (see https://github.com/llvm/llvm-project/issues/81286)
3. The current implementation of `__cxx_atomic_wait` is tightly coupled with `__cxx_atomic_impl`, which makes it impossible to use cleanly from `atomic_ref`. @huixie90 made the observation that "load" is a "customization point" for that operation, IOW it should be able to work with anything that can be loaded. @huixie90 will refactor that function to make it more generic, so that we can pass an `atomic_ref` to it out of the box.

dalg24 committed:
[libc++][atomic_ref] Fixup unsupported hardening mode fast

dalg24 committed:
[libc++][atomic_ref] revert hack to be able to use atomic wait implementation

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] Use __atomic_wait following Hui's refactor

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,335 @@
+  // whether the object pointed to has padding AND we will actually ignore them
+  static constexpr bool __might_have_padding =
+#  if __has_builtin(__builtin_clear_padding)
+      !has_unique_object_representations_v<_Tp> && !same_as<_Tp, float> && !same_as<_Tp, double>;
```
Ping?

huixie90 commented:
Hello sorry for the delay. I have merged the atomic wait refactoring and hopefully the wait and notify functions can be reused now.  Could you please rebase and add the atomic_watiable_traits for your type ? You can check atomic_base as an example

dalg24 commented:
> Hello sorry for the delay. I have merged the atomic wait refactoring and hopefully the wait and notify functions can be reused now. Could you please rebase and add the atomic_watiable_traits for your type ? You can check atomic_base as an example

I had already rebased on your work yesterday but I see they are some other conflicts in the meantime.  Will fix later today.

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] Suppress clang-tidy diagnostic when __builtin_clear_padding is not available

dalg24 committed:
[libc++][atomic_ref] Remove stray semicolon

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  atomic_ref& operator=(const atomic_ref&) = delete;
+
+  using difference_type = __base::value_type;
```
could you please put typedefs together?

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  atomic_ref& operator=(const atomic_ref&) = delete;
+
+  using difference_type = __base::value_type;
```
could you please put typedef on the top?

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  using difference_type = __base::value_type;
+
+  _LIBCPP_HIDE_FROM_ABI _Tp fetch_add(_Tp __arg, memory_order __order = memory_order_seq_cst) const noexcept {
```
not an issue. we do have builtins for RMW operations for floating points in clang. but i guess gcc does not have it?

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  atomic_ref& operator=(const atomic_ref&) = delete;
+
+  using difference_type = ptrdiff_t;
```
same as above

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+template <class _Tp>
+struct __atomic_ref_base {
+  _Tp* __ptr_;
+
+  _LIBCPP_HIDE_FROM_ABI static _Tp* __clear_padding(_Tp& __val) noexcept {
+    _Tp* __ptr = std::addressof(__val);
+#  if __has_builtin(__builtin_clear_padding)
+    __builtin_clear_padding(__ptr);
+#  endif
+    return __ptr;
+  }
```
could you please make them private/protected?

dalg24 replied:
Done

huixie90 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+    return __ptr;
+  }
+
+  _LIBCPP_HIDE_FROM_ABI __atomic_ref_base(_Tp& __obj) : __ptr_(std::addressof(__obj)) {}
```
perhaps put member functions below the typedefs and static member variables?

huixie90 commented on libcxx/test/std/atomics/atomics.ref/compare_exchange_weak.pass.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+void test() {
+  test_compare_exchange_weak<int>();
+
+  test_compare_exchange_weak<float>();
+
+  test_compare_exchange_weak<int*>();
+
+  struct X {
+    int i;
+    //X() = default;
+    X(int ii) noexcept : i(ii) {}
+    bool operator==(X o) const { return i == o.i; }
+  };
+  test_compare_exchange_weak<X>();
```
i think the tests are insufficient. 

We should test

1. user defined types
2. all integral types
3. all floating point types ( especially `long double` which is the most problematic one and we have lots of bugs at the moment)
4. pointer to all above types

this should apply to other tests too

dalg24 replied:
Done

huixie90 commented on libcxx/test/std/atomics/atomics.ref/load.pass.cpp:
```diff
@@ -0,0 +1,55 @@
+//
```
I think majority of the tests would still pass even if the implementation does not have any atomic operations. i would suggest to have a look at these tests
https://github.com/llvm/llvm-project/blob/main/libcxx/test/std/atomics/atomics.types.generic/atomics.types.float/test_helper.h

(most of the tests in atomic float uses the test template above)

dalg24 committed:
[libc++][atomic_ref] Mark P0019 as "Complete"

dalg24 committed:
[libc++][atomic_ref] Per review put typedefs together

ldionne commented on libcxx/docs/ReleaseNotes/19.rst:
```diff
@@ -44,6 +44,7 @@ Implemented Papers
 - P2495R3 - Interfacing ``stringstream``s with ``string_view``
 - P2302R4 - ``std::ranges::contains``
 - P1659R3 - ``std::ranges::starts_with`` and ``std::ranges::ends_with``
+- P0019R8 - ``atomic_ref``
```
```suggestion
- P0019R8 - ``std::atomic_ref``
```

ldionne commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  static constexpr size_t required_alignment = alignof(_Tp);
+
+  static constexpr bool is_always_lock_free =
```
Why not use `__libcpp_is_always_lock_free<_Tp>::value` here instead?

dalg24 replied:
See my comment above about `__atomic_always_lock_free(sizeof(_Tp), nullptr)` giving an error in the tests.

ldionne commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,345 @@
+
+  static constexpr size_t required_alignment = alignof(_Tp);
+
+  static constexpr bool is_always_lock_free =
+      __atomic_always_lock_free(sizeof(_Tp), reinterpret_cast<void*>(-required_alignment));
+
+  _LIBCPP_HIDE_FROM_ABI bool is_lock_free() const noexcept {
+    return __atomic_is_lock_free(sizeof(_Tp), reinterpret_cast<void*>(-required_alignment));
+  }
```
In fact why not use `__cxx_atomic_is_lock_free(sizeof(_Tp))` like we do for `atomic<_Tp>`?

dalg24 replied:
@huixie90 The test
```C++
    if (std::atomic_ref<T>::is_always_lock_free) {
      std::same_as<bool> auto is_lock_free = a.is_lock_free();
      assert(is_lock_free);
    }
```
was failing with `__atomic_always_lock_free(sizeof(_Tp), nullptr)` so I had gone for a fake, minimally aligned pointer.
I proposed to use ` __builtin_align_up(static_cast<_Tp*>(nullptr), required_alignment)` and properly passed the `__ptr_` data member to `__atomic_lock_free` instead.  Let me know what you think.

dalg24 committed:
[libc++][atomic_ref] Fixup release note

dalg24 reviewed (commented):
Partial answer to Hui's review

dalg24 committed:
[libc++][atomic_ref] Propose new is_[always_]lock_free implementation w/o fake minimally aligned ptr

dalg24 committed:
[libc++][atomic_ref] GCC does not provide __builtin_align_up

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,343 @@
+  using difference_type = __base::value_type;
+
+  _LIBCPP_HIDE_FROM_ABI explicit atomic_ref(_Tp& __obj) : __base(__obj) {
+    _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN((uintptr_t)addressof(__obj) % __base::required_alignment == 0,
```
Done

dalg24 committed:
[libc++][atomic_ref] Prefer reinterpret_cast to C-style cast and qualify addressof calls in constructor precondition checks

dalg24 committed:
[libc++][atomic_ref] Add comment suggested by Louis to describe how the compare exchange with clear padding works

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,343 @@
+    ) {
+      return __atomic_compare_exchange(__ptr, __expected, __desired, __is_weak, __success, __failure);
+    } else { // _Tp has padding bits and __builtin_clear_padding is available
+      __clear_padding(*__desired);
```
Added

dalg24 committed:
[libc++][atomic_ref] Use protected access specifier in __atomic_ref_base

dalg24 commented on libcxx/docs/Status/Cxx20Papers.csv:
```diff
@@ -26,7 +26,7 @@
 "`P0966R1 <https://wg21.link/P0966R1>`__","LWG","``string::reserve``\  Should Not Shrink","Jacksonville","|Complete| [#note-P0966]_","12.0"
 "","","","","","",""
-"`P0019R8 <https://wg21.link/P0019R8>`__","LWG","Atomic Ref","Rapperswil","",""
+"`P0019R8 <https://wg21.link/P0019R8>`__","LWG","Atomic Ref","Rapperswil","[Complete]",""
```
Done

dalg24 committed:
[libc++][atomic_ref] Specify release in cxx20 status table

dalg24 committed:
[libc++][atomic_ref] __clear_padding and __compare_exchange can be private

dalg24 committed:
[libc++][atomic_ref] Prefer std::byte for raw memory buffers

dalg24 commented on libcxx/include/__atomic/atomic_ref.h:
```diff
@@ -0,0 +1,343 @@
+        __order == memory_order::relaxed || __order == memory_order::consume || __order == memory_order::acquire ||
+            __order == memory_order::seq_cst,
+        "atomic_ref: memory order argument to atomic load operation is invalid");
+    alignas(_Tp) unsigned char __mem[sizeof(_Tp)];
```
Sure.  Done.

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] Instantiate tests for more types

dalg24 committed:
[libc++][atomic_ref] Instantiate libcxx tests for more types

dalg24 committed:
[libc++][atomic_ref] assert integral type

dalg24 committed:
[libc++][atomic_ref] casually disable long double tests that failed

dalg24 committed:
[libc++][atomic_ref] assert integral type in xor

dalg24 committed:
[libc++][atomic_ref] forgot fetch_sub

dalg24 committed:
[libc++][atomic_ref] Add tests that actually check the aromicity

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] delete trailing empty line in test helper that clang-format didnt like

dalg24 committed:
[libc++][atomic_ref] fix redeclaration of x shadow warning

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] Mark notify_{all,one}/wait test as unsupported with no threads

EricWF commented on libcxx/test/std/atomics/atomics.ref/notify_all.pass.cpp:
```diff
@@ -6,6 +6,7 @@
 //===----------------------------------------------------------------------===//
 
 // UNSUPPORTED: c++03, c++11, c++14, c++17
+// UNSUPPORTED: no-threads
```
What occurs when you run these tests at the moment?

Do they hang forever? Do we expect them to fail (and should consider `XFAIL`)?

dalg24 replied:
Does not compile
https://github.com/llvm/llvm-project/actions/runs/8939256898/job/24556637589#step:3:1864

I am following what is done in the `std::atomic` tests
https://github.com/llvm/llvm-project/blob/5d81b1c50a6eadc5d2ebe8c524dd2daa665402d7/libcxx/test/std/atomics/atomics.types.generic/atomics.types.float/notify_all.pass.cpp#L8

mordante replied:
This test and several others fail on apple https://buildkite.com/llvm-project/libcxx-ci/builds/35217#_. These platforms lack synchronization. This should fix these builds.
```suggestion
// XFAIL: availability-synchronization_library-missing

```

mordante commented on libcxx/test/std/atomics/atomics.ref/fetch_add.pass.cpp:
```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14, c++17
+
```
For the failures on picolib https://buildkite.com/llvm-project/libcxx-ci/builds/35217#018f401d-0d82-4d0a-8539-22802e30de78  I expect this fixes it.
```suggestion
// UNSUPPORTED: libcpp-has-no-threads

```

dalg24 replied:
Done.  Thank you.

mordante reviewed (commented):
Some proposed CI fixes.

ldionne commented on libcxx/test/libcxx/atomics/atomics.ref/assert.ctor.pass.cpp:
```diff
@@ -0,0 +1,38 @@
+
+  TEST_LIBCPP_ASSERT_FAILURE(
+      ([] {
+        char c[8];
```
```suggestion
        alignas(float) std::byte c[2 * sizeof(float)]; // intentionally larger
```

ldionne commented on libcxx/test/libcxx/atomics/atomics.ref/assert.ctor.pass.cpp:
```diff
@@ -0,0 +1,38 @@
+  TEST_LIBCPP_ASSERT_FAILURE(
+      ([] {
+        char c[8];
+        float* f = new (c + 1) float(3.14f);
```
```suggestion
        float* f = new (c + 1) float(3.14f); // intentionally misaligned
```

ldionne commented on libcxx/test/std/atomics/atomics.ref/assign.pass.cpp:
```diff
@@ -0,0 +1,54 @@
+    T x(T(1));
+    std::atomic_ref<T> const a(x);
+
+    a = T(2);
```
You should test that the return type of `operator=` is `T`, and that it returns the right value.

```
decltype(auto) std::same_as<T> result = (a = T(2));
assert(result == T(2));
```

ldionne reviewed (approved):
This basically LGTM once CI is all passing. Please don't hesitate to ping me on Discord if you need help with specific CI failures and how they should map to `UNSUPPORTED` and/or `XFAIL`. This is tricky to get right and atomic is full of them.

Thanks for the patch, there was a lot of work needed to get it into its current state but I think this is really good!

dalg24 committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into atomic_ref

dalg24 committed:
[libc++][atomic_ref] Fixup AtomicRef status entry

dalg24 committed:
[libc++][atomic_ref] Also mark P1643R1 (add wait/notify to atomic_ref) as complete

dalg24 committed:
[libc++][atomic_ref] Add empty line before the first REQUIRES

dalg24 committed:
[libc++][atomic_ref] Add comment where no assertion is expected to fail

dalg24 committed:
[libc++][atomic_ref] Prefer TestEachAtomicType

dalg24 committed:
[libc++][atomic_ref] Improve misaligned objects test

dalg24 committed:
[libc++][atomic_ref] Add a comment for the definition of is_always_lock_free

dalg24 committed:
[libc++][atomic_ref] Fix typo memory_or[e]der

dalg24 committed:
[libc++][atomic_ref] Inline test() into main and use TestEachAtomicType

dalg24 committed:
[libc++][atomic_ref] check return type with same_as decltype(auto) idiom

dalg24 committed:
[libc++][atomic_ref] improve operator&= test

dalg24 committed:
[libc++][atomic_ref] Fix typo in noexecpt check for operator|=

dalg24 committed:
[libc++][atomic_ref] Add missing <concepts> header include

dalg24 committed:
[libc++][atomic_ref] Fix shadowed declaration warning

dalg24 committed:
[libc++][atomic_ref] Fix silly redefinition of a buffer of raw unitialized memory

dalg24 committed:
[libc++][atomic_ref] Add XFAIL: !has-64-bit-atomics

dalg24 committed:
[libc++][atomic_ref] Add XFAIL: !has-64-bit-atomics and XFAIL: availability-synchronization_library-missing for wait/notify

dalg24 committed:
[libc++][atomic_ref] Fixup fetch_{add,sub} do not use TestEachFloatingPointType

dalg24 committed:
[libc++][atomic_ref] Add XFAIL: !has-1024-bit-atomics to "fix" the windows builds

dalg24 committed:
[libc++][atomic_ref] Attempt to drop XFAIL: !has-64-bit-atomics in a couple places to see if it fixes the Armv7-M picolibc builds

dalg24 committed:
Merge branch 'main' into atomic_ref
